session.go

```diff
@@ -41,6 +42,59 @@ func SessionJoin(app *App) func(c echo.Context) error {
 	}
 }
 
+// /game/joinserver.jsp
+func SessionJoinServer(app *App) func(c echo.Context) error {
+	return func(c echo.Context) error {
+		username := c.QueryParam("user")
+		sessionID := c.QueryParam("sessionId")
+		serverID := c.QueryParam("serverId")
+
+		// If any parameters are missing, return NO
+		if username == "" || sessionID == "" || serverID == "" {
+			return c.String(http.StatusMethodNotAllowed, "NO")
+		}
+
+		// Parse sessionId. It has the form:
+		// token:<accessToken>:<player UUID>
+		split := strings.Split(sessionID, ":")
+		if len(split) != 3 || split[0] != "token" {
+			return c.String(http.StatusMethodNotAllowed, "NO")
+		}
+		accessToken := split[1]
+		id := split[2]
+
+		// Is the accessToken valid?
+		client := app.GetClient(accessToken, StalePolicyDeny)
+		if client == nil {
+			return c.String(http.StatusMethodNotAllowed, "NO")
+		}
+
+		// If the player name corresponding to the access token doesn't match
+		// the `user` param from the request, return NO
+		user := client.User
+		if user.PlayerName != username {
+			return c.String(http.StatusMethodNotAllowed, "NO")
+		}
+		// If the player's UUID doesn't match the UUID in the sessionId, return
+		// NO
+		userID, err := UUIDToID(user.UUID)
+		if err != nil {
+			return err
+		}
+		if userID != id {
+			return c.String(http.StatusMethodNotAllowed, "NO")
+		}
+
+		user.ServerID = MakeNullString(&serverID)
+		result := app.DB.Save(&user)
+		if result.Error != nil {
+			return result.Error
+		}
+
+		return c.String(http.StatusOK, "YES")
+	}
```

Comment:
This is correct, however it's missing the fallback API, you have 2 choices:
- Take the proper (and harder) path, which is to mimic modern clients private and pubkeys and serverhash for sessionJoin
- Use the still working legacy domain https://session.minecraft.net/game/joinserver.jsp

I opted for 2nd option, but other fallback APIs like ely.by use /legacy/ as an endpoint: [Source](https://docs.ely.by/en/minecraft-auth.html?highlight=legacy#id5) so it might be worth taking into consideration

(If only things were standardised...)

Reply:
Also, since this **joinserver** and not checkserver, "OK" is the correct answer and **not** "YES", the "NO" part can be anything like an error (see the code snippets i sent in matrix for more context)

Reply:
About the 2nd option, reusing the same code for /session/hasJoined while adapting it for the join function worked, here is the snippet if you're interested, but i'm not sure this may be the right way of doing it, so consider it a "hack"
```go
		// Is the accessToken valid?
		client := app.GetClient(accessToken, StalePolicyDeny)
		if client == nil {

			// If invalid, check the fallback servers
			for _, fallbackAPIServer := range app.Config.FallbackAPIServers {
				base, err := url.Parse(fallbackAPIServer.SessionURL)
				if err != nil {
					log.Println(err)
					continue
				}
				// replace string sessionserver.mojang.com with session.minecraft.net in the URL
				if strings.Contains(base.Host, "sessionserver.mojang.com") {
					base.Host = strings.ReplaceAll(base.Host, "sessionserver.mojang.com", "session.minecraft.net")
				}

				base.Path += "/game/joinserver.jsp"
				params := url.Values{}
				params.Add("user", username)
				params.Add("sessionId", sessionID)
				params.Add("serverId", serverID)
				base.RawQuery = params.Encode()

				res, err := MakeHTTPClient().Get(base.String())
				if err != nil {
					log.Printf("Received invalid response from fallback API server at %s\n", base.String())
					continue
				}
				defer res.Body.Close()

				// Check the body, if it contains "OK", return "OK"
				// Otherwise, return an error message to the player.
				body, err := io.ReadAll(res.Body) // Read the body
				if err != nil {
					return err
				}


				if res.StatusCode == http.StatusOK && string(body) == "OK" {
					return c.String(http.StatusOK, "OK")
				}
			}
			return c.String(http.StatusOK, "Invalid access token. Try restarting the game.")
		}

```

Reply:
Ah, I think I understand what you mean about the fallbackAPIServer functionality. The `/session/minecraft/join` and `/game/joinserver.jsp` routes should never/don't ever need to query fallback API servers. These routes are called by the Minecraft client when connecting to a Minecraft server. The client knows an `accessToken` provided by the launcher as well as a URL for an authentication/session server. This URL can be either provided by the launcher, hardcoded/patched in, patched by a Java agent, redirected by a proxy, or some combination of those. The client's `accessToken` **MUST** be issued by the session server pointed to by that URL. Otherwise, the client would send an `accessToken` issued by server A to server B, and server B would be able to impersonate the client to server A.

So we never need to query fallback API servers for the "session join" routes. If everything is set up sanely, the `accessToken` WILL be issued from the same authentication server that gets the `join` request. If your client has a mismatched auth server URL and accessToken, then something has gone horribly wrong.

Reply:
I think I finally understand what you mean, I just realised how critical it was, a vanilla minecraft server *directly* authenticates through mojang's sessionserver domains, we only have to check from the serverside if the user *actually* joined or not, good point, my testing confused me as i kept thinking i was supposed to proxy the login part too to access from a mojang account through a proxied minecraft instance that redirects everything through drasl's servers. In the end yes, this does not need to be sent to drasl and can be disregarded. 
Thanks for the insight! This can be disregarded.